dev

danya1705 opened the pull request.

danya1705 committed:
Initial project settings, DTO classes added

danya1705 committed:
Исправлены DTO классы согласно обновлённому OpenAPI

Modusen committed:
Добавлены контроллеры и методы с описанием и аннотациями. Практически без тел методов. Привел к виду по ТЗ в сваггере

Barsev174 committed:
Добавлены контроллеры и методы с заглушками для работы с объявлениями

IGuruI-1981 committed:
Добавлены эндпоинты в AdsController для комментариев к объявлениям.

IGuruI-1981 committed:
Скорректированы таги в обявлениях

Barsev174 committed:
Поправил форматирование и тэги

dmitry-bizin commented on src/main/java/ru/skypro/homework/WebSecurityConfig.java:
```diff
@@ -45,8 +45,9 @@ public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
+//                    .authenticated()
+                    )
         .cors()
         .disable()
```
здесь вместо disable() нужно and()

dmitry-bizin commented on src/main/java/ru/skypro/homework/controller/AdsController.java:
```diff
@@ -0,0 +1,288 @@
+                    )})
+    })
+    @PostMapping
+    public AdsDto addAd(@RequestBody AdsDto adsDto) {
```
в спецификации немного не так. Там во-первых отправляется другой тип запроса, а во-вторых в запросе сразу картинка и json

dmitry-bizin commented on src/main/java/ru/skypro/homework/controller/AdsController.java:
```diff
@@ -0,0 +1,288 @@
+                    )})
+    })
+    @PatchMapping("/{id}")
+    public AdsDto updateAds(@PathVariable Long id) {
```
а что обновляется-то?)

dmitry-bizin commented on src/main/java/ru/skypro/homework/controller/AdsController.java:
```diff
@@ -0,0 +1,288 @@
+                    )})
+    })
+    @PatchMapping("/id/image")
+    public AdsDto updateImage(@PathVariable Long id) {
```
а где картинка?

dmitry-bizin commented on src/main/java/ru/skypro/homework/controller/AdsController.java:
```diff
@@ -0,0 +1,288 @@
+                    )})
+    })
+    @GetMapping("/{id}/comments")
+    public CommentDto getCommentsAds(@RequestBody CommentDto commentDto) {
```
get метод не может иметь тела

dmitry-bizin commented on src/main/java/ru/skypro/homework/controller/AdsController.java:
```diff
@@ -0,0 +1,288 @@
+                    )})
+    })
+    @DeleteMapping("/{adId}/comments/{commentId}")
+    public CommentDto deleteCommentAds(@RequestBody CommentDto commentDto) {
```
зачем при удалении отправлять в теле комментарий? в спецификации не так

dmitry-bizin commented on src/main/java/ru/skypro/homework/controller/UserController.java:
```diff
@@ -0,0 +1,102 @@
+                    content = @Content)
+    })
+    @GetMapping("/me")
+    public UserDto getUser(@RequestBody UserDto user) {
```
у get методов нет тела запроса

dmitry-bizin commented on src/main/java/ru/skypro/homework/controller/UserController.java:
```diff
@@ -0,0 +1,102 @@
+                    description = "Not Found")
+    })
+    @PatchMapping(value = "/me/image", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
+    public void updateUserImage(@RequestParam MultipartFile avatar) {
```
название части неверное (см. спецификацию)

dmitry-bizin commented on src/main/java/ru/skypro/homework/dto/ResponseWrapperAdsDto.java:
```diff
@@ -0,0 +1,9 @@
+@Data
+public class ResponseWrapperAdsDto {
+    private int count;
+    private AdsDto[] results;
```
а почему массив? со списком удобнее работать

dmitry-bizin commented on src/main/java/ru/skypro/homework/dto/ResponseWrapperCommentDto.java:
```diff
@@ -0,0 +1,9 @@
+@Data
+public class ResponseWrapperCommentDto {
+    private int count;
+    private CommentDto[] results;
```
а почему массив? со списком удобнее работать

danya1705 committed:
Заменены массивы на списки в полях двух DTO

Modusen committed:
Поправлены комментарии ПР в юзер контроллере.

Modusen committed:
Поправлены комментарии ПР в юзер контроллере.

Modusen committed:
Merge branch 'dmitriy-feature' into dev

Barsev174 committed:
Поправил согласно замечаниям наставника

Barsev174 committed:
Merge remote-tracking branch 'origin/dev' into dev

IGuruI-1981 committed:
Эндпоинты комментариев перенесены в отдельный контроллер .

IGuruI-1981 committed:
Эндпоинты комментариев перенесены в отдельный контроллер .

IGuruI-1981 committed:
Merge branch 'viktor-feature' into dev

IGuruI-1981 committed:
Эндпоинты комментариев перенесены в отдельный контроллер .

IGuruI-1981 committed:
Эндпоинты комментариев перенесены в отдельный контроллер .

IGuruI-1981 committed:
Merge branch 'viktor-feature' into dev

danya1705 committed:
Merge remote-tracking branch 'origin/dev' into dev

IGuruI-1981 committed:
Добавлены сущности, репозитории, связь с БД, зависимости

IGuruI-1981 committed:
Merge branch 'viktor-feature' into dev

danya1705 committed:
Merge remote-tracking branch 'origin/dev' into dev

IGuruI-1981 committed:
Добавлены конструкторы, геттеры и сеттеры

IGuruI-1981 committed:
Merge branch 'viktor-feature' into dev

IGuruI-1981 committed:
Внесены изменения в App.properties

IGuruI-1981 committed:
Добалена аннотация Data

IGuruI-1981 committed:
Merge branch 'viktor-feature' into dev

IGuruI-1981 committed:
refactor

IGuruI-1981 committed:
Merge branch 'viktor-feature' into dev

danya1705 committed:
Merge remote-tracking branch 'origin/dev' into dev

IGuruI-1981 committed:
refactor

IGuruI-1981 committed:
Merge branch 'viktor-feature' into dev

danya1705 committed:
Merge remote-tracking branch 'origin/dev' into dev

IGuruI-1981 committed:
refactor 2

IGuruI-1981 committed:
Merge branch 'viktor-feature' into dev

danya1705 committed:
Merge remote-tracking branch 'origin/dev' into dev

dmitry-bizin commented on src/main/java/ru/skypro/homework/controller/AdsController.java:
```diff
@@ -0,0 +1,193 @@
+                    )})
+    })
+    @PostMapping
+    public AdsDto addAd(@RequestBody CreateAdsDto createAdsDto,@RequestBody String image) {
```
аннотации неверные снова, этот метод не будет работать. Я на разборе рассказывал и показывал, какие должны быть аннотации, рассказывал про multipart-запросы

dmitry-bizin commented on src/main/java/ru/skypro/homework/controller/AdsController.java:
```diff
@@ -0,0 +1,193 @@
+                    )})
+    })
+    @PatchMapping("/{id}/image")
+    public AdsDto updateImage(@PathVariable Long id, @RequestBody String image) {
```
здесь тоже неверно. Как вы собираетесь картинку передавать в виде строки? Рекомендую посмотреть домашку по аватаркам, здесь будет всё то же самое, ничего нового.

dmitry-bizin commented on src/main/java/ru/skypro/homework/entity/Ads.java:
```diff
@@ -0,0 +1,31 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Integer pk;
+
+    private Integer author;
```
а где связь с таблицей юзеров?

dmitry-bizin commented on src/main/java/ru/skypro/homework/entity/Ads.java:
```diff
@@ -0,0 +1,31 @@
+
+    private int price ;
+
+    private String firstName;
```
аномалия БД - у вас дублируются поля. Здесь должна быть связь на юзера, сущность != dto, поля будут разные

dmitry-bizin commented on src/main/java/ru/skypro/homework/entity/Comment.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Id
+
+    private Integer adId;
```
гле связь с объявлением?

dmitry-bizin commented on src/main/java/ru/skypro/homework/entity/Comment.java:
```diff
@@ -0,0 +1,27 @@
+    @Id
+    private Integer commentId;
+
+    private Integer author;
```
где связь с юзером?

dmitry-bizin commented on src/main/java/ru/skypro/homework/entity/Ads.java:
```diff
@@ -0,0 +1,31 @@
+
+    private String description;
+
+    private String image;
```
как картинка у вас будет храниться? что такое строка?

dmitry-bizin commented on src/main/resources/liquibase/sql/user.sql:
```diff
@@ -0,0 +1,59 @@
+
+    price INTEGER NOT NULL ,
+
+    first_name TEXT NOT NULL ,
```
аномалия - дублирование данных

danya1705 committed:
Добавлена сущность Image для хранения данных о картинках;
Настроены все связи между сущностями через Гибернет;
Исправлены эндпойнты addAd и updateImage в AdsController;

danya1705 committed:
Создана миграция в liquibase:
1) удален столбец first_name из таблицы ads
2) создана таблица image

danya1705 committed:
Создана миграция в liquibase: заменены типы полей image в таблицах user_ads и ads с text на integer, и им назначен foreign key на image.id

danya1705 committed:
В pom.xml добавлены зависимости и плагины Mapstruct

danya1705 committed:
В сущность Ads добавленое связанное поле commentsList.
Добавлены классы UserService и AdsService с методами getById в каждом из них.
Созданы маперы для комментариев и массивов комментариев.
Написаны тесты для созданных маперов.

danya1705 committed:
Переделал CommentListMapper на интерфейс

Barsev174 committed:
Добавил маперы для работы с объявлениями

Modusen committed:
Добавлен маппер на Юзера. Написаны тесты на маппер. Тесты проходятся успешно.

danya1705 committed:
Исправлен AdsMapper
Написаны тесты для AdsMapper и AdsListMapper
UserMapper сделан интерфейсом

dmitry-bizin commented on pom.xml:
```diff
@@ -100,8 +104,8 @@
         <configuration>
-          <source>1.8</source>
-          <target>1.8</target>
+          <source>9</source>
```
а почему 9 версия Java? у вас же 11 выше указана

dmitry-bizin commented on src/main/java/ru/skypro/homework/mapper/AdsMapper.java:
```diff
@@ -0,0 +1,31 @@
+
+    @Mapping(source = "author.firstName", target = "authorFirstName")
+    @Mapping(source = "author.lastName", target = "authorLastName")
+    @Mapping(source = "image.filePath", target = "image")
```
в image не путь в файловой системе должен быть, а ссылка на получение картинки

dmitry-bizin commented on src/main/java/ru/skypro/homework/mapper/AdsMapper.java:
```diff
@@ -0,0 +1,31 @@
+public interface AdsMapper {
+
+    @Mapping(source = "author.id", target = "author")
+    @Mapping(source = "image.filePath", target = "image")
```
в image не путь в файловой системе должен быть, а ссылка на получение картинки

dmitry-bizin commented on src/main/java/ru/skypro/homework/mapper/CommentMapper.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Mapping(source = "commentId", target = "pk")
+    @Mapping(source = "author.id", target = "author")
+    @Mapping(source = "author.image.filePath", target = "authorImage")
```
в image не путь в файловой системе должен быть, а ссылка на получение картинки

dmitry-bizin commented on src/main/java/ru/skypro/homework/mapper/CommentMapper.java:
```diff
@@ -0,0 +1,46 @@
+@Mapper(componentModel = "spring", injectionStrategy = InjectionStrategy.CONSTRUCTOR)
+public abstract class CommentMapper {
+
+    @Autowired
```
инъекцию лучше делать через конструктор

dmitry-bizin commented on src/main/java/ru/skypro/homework/mapper/CommentMapper.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Autowired
+    private UserService userService;
+    @Autowired
```
инъекцию лучше делать через конструктор

dmitry-bizin commented on src/main/java/ru/skypro/homework/mapper/UserMapper.java:
```diff
@@ -0,0 +1,27 @@
+    @Mapping(source = "user.lastName", target = "lastName")
+    @Mapping(source = "user.phone", target = "phone")
+    @Mapping(source = "user.email", target = "email")
+    @Mapping(source = "user.image.filePath", target = "image")
```
в image не путь в файловой системе должен быть, а ссылка на получение картинки

dmitry-bizin commented on src/main/java/ru/skypro/homework/mapper/UserMapper.java:
```diff
@@ -0,0 +1,27 @@
+    @Mapping(source = "user.image.filePath", target = "image")
+    UserDto toDto(User user);
+
+    @Mapping(source = "userDto.image", target = "user.image.filePath")
```
в image не путь в файловой системе должен быть, а ссылка на получение картинки

danya1705 committed:
В сущность Image добавлено поле "url" с прямыми ссылками на картинки, cделана соответствующая миграция, в мапперах заменено поле Image.filePath на Image.url и поправлены их тесты.
Добавлена конфигурация MvcConfig с настройкой путей к файлам.
Созданы ImageRepository и ImageService для загрузки картинок.
Для UserController.updateUserImage, AdsController.addAd и AdsController.updateImage сделаны сервисы.

danya1705 committed:
Добавлены контроллер, сервис и метод репозитория для поиска объявлений по названию.
Исправлен AdsListMapper.

IGuruI-1981 committed:
Добавлен CommentController и методы

Modusen committed:
Написан сервис для User'a, косметические изменения User контроллера

IGuruI-1981 committed:
Merge branch 'dev' into viktor-feature

Barsev174 committed:
Добавил методы для работы с объявлениями

IGuruI-1981 committed:
Merge branch 'dev' into viktor-feature

IGuruI-1981 committed:
Добавлены Сервисы комментов

danya1705 committed:
В сущности Comment составной ключ заменён на простой, сделана соответствующая миграция в liquibase.
В CommentController поправлены коды ответов в описаниях.
В CommentService добавлен метод updateComment и проверки для Optinal.
CommentMapper сделан интерфейсом.

IGuruI-1981 committed:
Удалена сщность CommentId.java

danya1705 committed:
Поправлены AdsService,AdsController и AdsRepository

dmitry-bizin commented on src/main/java/ru/skypro/homework/controller/UserController.java:
```diff
@@ -5,97 +5,92 @@
-        return new UserDto();
+    public ResponseEntity<UserDto> getUser() {
+        if (userService.getUserInfo(1) != null) {
+            return ResponseEntity.ok(userService.getUserInfo(1));
```
не очень хорошо 2 раза выполнять метод. Надо выполнить 1 раз, запомнить результат и далее работать с результатом

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/ImageService.java:
```diff
@@ -0,0 +1,55 @@
+        Files.deleteIfExists(filePath);
+        try (
+                InputStream is = imageFile.getInputStream();
+                OutputStream os = Files.newOutputStream(filePath, CREATE_NEW);
```
запись массива байт в файл можно сделать одной строкой через Files.write

dmitry-bizin commented on src/main/java/ru/skypro/homework/service/ImageService.java:
```diff
@@ -0,0 +1,55 @@
+        image.setFilePath(filePath.toString());
+        image.setFileSize(imageFile.getSize());
+        image.setMediaType(imageFile.getContentType());
+        image.setUrl(serverAddress + ":" + serverPort + "/img/" + imageFile.getOriginalFilename());
```
вот здесь дублирование информации по идее. Зачем хранить url? а вдруг он поменяется? лучше его налету просто формировать в мапперах

dmitry-bizin replied:
при смене хоста, урла, порта вам придётся писать миграцию данных, что не очень хорошо. К тому же, это просто дублирование информации, т.к. у вас в filePath есть инфа, где лежит файл

danya1705 committed:
Внесены исправления по замечаниям к дз

danya1705 committed:
Настроена аутентификация

danya1705 committed:
Настроена авторизация, внесены правки в контроллеры, сервисы и мапперы

BizinMitya commented on src/main/java/ru/skypro/homework/config/WebSecurityConfig.java:
```diff
@@ -0,0 +1,64 @@
+            "/v3/api-docs",
+            "/webjars/**",
+            "/login",
+            "/register"
```
здесь должен быть ещё /ads эндпоинт, потому что по заданию все объявления можно получить без авторизации

dmitry-bizin commented on src/main/java/ru/skypro/homework/controller/AdsController.java:
```diff
@@ -80,9 +81,11 @@ public ResponseEntity<FullAdsDto> getAd(@PathVariable Integer id) {
+    public ResponseEntity<?> deleteAd(@PathVariable Integer id, Authentication authentication) {
+        if (adsService.deleteAd(id, authentication.getName())) {
+            return ResponseEntity.ok().build();
+        } else
```
фигурные скобки

dmitry-bizin commented on src/main/java/ru/skypro/homework/mapper/AdsMapper.java:
```diff
@@ -15,15 +15,17 @@
     @Mapping(source = "author.id", target = "author")
-    @Mapping(source = "image.url", target = "image")
+    @Mapping(target = "image",
+            expression = "java(\"/img/\" + java.nio.file.Path.of(ad.getImage().getFilePath()).getFileName().toString())")
```
это неверно. Здесь надо передавать эндпоинт, а не путь к файлу в файловой системе

dmitry-bizin commented on src/main/java/ru/skypro/homework/mapper/AdsMapper.java:
```diff
@@ -15,15 +15,17 @@
     @Mapping(source = "author.lastName", target = "authorLastName")
-    @Mapping(source = "image.url", target = "image")
+    @Mapping(target = "image",
+            expression = "java(\"/img/\" + java.nio.file.Path.of(ad.getImage().getFilePath()).getFileName().toString())")
```
это неверно. Здесь надо передавать эндпоинт, а не путь к файлу в файловой системе

dmitry-bizin commented on src/main/java/ru/skypro/homework/mapper/CommentMapper.java:
```diff
@@ -13,7 +13,8 @@ public interface CommentMapper {
     @Mapping(source = "author.id", target = "author")
-    @Mapping(source = "author.image.url", target = "authorImage")
+    @Mapping(target = "authorImage",
+            expression = "java(\"/img/\" + java.nio.file.Path.of(comment.getAuthor().getImage().getFilePath()).getFileName().toString())")
```
это неверно. Здесь надо передавать эндпоинт, а не путь к файлу в файловой системе

dmitry-bizin commented on src/main/java/ru/skypro/homework/mapper/UserMapper.java:
```diff
@@ -16,12 +20,41 @@ public interface UserMapper {
-    @Mapping(source = "user.image.url", target = "image")
-    UserDto toDto(User user);
+    @Mapping(target = "image",
+            expression = "java(\"/img/\" + java.nio.file.Path.of(user.getImage().getFilePath()).getFileName().toString())")
```
это неверно. Здесь надо передавать эндпоинт, а не путь к файлу в файловой системе

dmitry-bizin commented on src/main/java/ru/skypro/homework/mapper/UserMapper.java:
```diff
@@ -16,12 +20,41 @@ public interface UserMapper {
-    @Mapping(source = "userDto.image", target = "user.image.url")
+        Image image = user.getImage();
+        if (image != null) {
+            userDto.setImage("/img/" + Path.of(image.getFilePath()).getFileName().toString());
```
это неверно. Здесь надо передавать эндпоинт, а не путь к файлу в файловой системе

dmitry-bizin commented on src/main/java/ru/skypro/homework/security/UserDetailsImpl.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return null;
```
почему null? надо брать роль у юзера и возвращать её

danya1705 committed:
Сделаны интегральные тесты на AdsController

danya1705 committed:
Сделаны интегральные тесты на AuthController

Barsev174 committed:
Отредактировал README.md

Modusen committed:
Добавлен эндпоинт на получение картинки.
Обновлен маппер

Modusen committed:
Merge remote-tracking branch 'origin/dev' into dev

danya1705 committed:
Эндпойнт для получения картинки перенесён в отдельный контроллер.
Сделаны тесты CommentController, UserController и ImageController.
Поправлены тесты мапперов с учетом внесённых в код изменений.
Код почищен от лишних импортов, закомментированных кусков и неиспользуемых методов.

danya1705 committed:
Добавлена передача роли в UserDetails

dmitry-bizin commented on src/main/java/ru/skypro/homework/config/WebSecurityConfig.java:
```diff
@@ -48,7 +35,7 @@ public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
                                         .mvcMatchers(HttpMethod.GET, "/ads")
                                         .permitAll()
-                                        .mvcMatchers("/ads/**", "/users/**")
+                                        .mvcMatchers("/ads/**", "/users/**", "/image/**")
```
хм... а как у вас будут картинки объявлений показываться на главной страницы до логина? Там авторизация не нужна. И для аватарок не нужна, т.к. они показываются в комментариях и там не нужна авторизация для их получения (авторизация именно того, чья аватарка)

dmitry-bizin replied:
есть сомнения, что это работает)

dmitry-bizin commented on src/main/java/ru/skypro/homework/config/MvcConfig.java:
```diff
@@ -17,6 +17,5 @@ public class MvcConfig implements WebMvcConfigurer {
     public final void addResourceHandlers(final ResourceHandlerRegistry registry) {
```
этот конфиг тогда не нужен, раз вы переделали?

Modusen committed:
JavaDoc комментарии

danya1705 committed:
Переделано получение всех картинок через эндпойнт, эндпойнту назначен доступ без аутентификации.
Добавлено удаление старой картинки из базы и файловой системы при редактировании изображений пользователя или объявления, а также при удалении объявления целиком.
Сделано переименование картинок при загрузке через генерацию случайного имени.

danya1705 committed:
Поправлены тесты контроллеров и мапперов под новую логику работы с картинками.

Barsev174 committed:
Отредактировал README.md

danya1705 committed:
Удалён MvcConfig

danya1705 committed:
Подредактирован JavaDoc

danya1705 committed:
Исправлена ошибка с пользователем без аватара.
